caddyhttp/fastcgi/fastcgi.go

```diff
@@ -91,14 +124,18 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error)
 				resp, err = fcgiBackend.Post(env, r.Method, r.Header.Get("Content-Type"), r.Body, contentLength)
 			}
 
-			if resp.Body != nil {
-				defer resp.Body.Close()
-			}
-
 			if err != nil && err != io.EOF {
+				persistentConnections[network+address].Close()
+				persistentConnections[network+address] = nil
```

Comment:
access to the persistentConnections is locked via poorMansSerialisation. But you are right, I also need a lock for creating the second lists entries. Thanks!